d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -123,24 +135,34 @@ public void registerXdsConnectionListener(XdsConnectionListener listener)
 
   public void listenToCluster(String clusterName)
   {
-    _watchedClusterResources.computeIfAbsent(clusterName, k ->
+    // if cluster name is a symlink, watch for D2SymlinkNode instead
+    String resourceName = D2_CLUSTER_NODE_PREFIX + clusterName;
+    if (shouldNotListenToSymlink(clusterName, resourceName))
```

Comment:
Confusing, `shouldNotListen...` being true shouldn't go to the if block. This can be done:
```
if (isSymlink)
    listen to symlink
else
    listen to cluster
```
Same for URIs.

Reply:
right, but the "listen to symlink" logic is the same for cluster and uri, so I put it into a helper function. I'll separate the isSymlink and the listenToSymlink to two functions then.